modules/core/src/minmax.cpp

```diff
@@ -71,33 +71,815 @@ minMaxIdx_( const T* src, const uchar* mask, WT* _minVal, WT* _maxVal,
     *_maxVal = maxVal;
 }
 
+#if CV_SIMD128
+#ifdef _MSC_VER
+    #define forceinline __forceinline
+#elif defined(__GNUC__)
+    #define forceinline inline __attribute__((__always_inline__))
+#elif defined(__CLANG__)
+    #if __has_attribute(__always_inline__)
+        #define forceinline inline __attribute__((__always_inline__))
+    #else
+        #define forceinline inline
+    #endif
+#else
+    #define forceinline inline
+#endif
+
+template<typename T, typename WT> forceinline void
+minMaxIdx_init( const T* src, const uchar* mask, WT* minval, WT* maxval,
+                size_t* minidx, size_t* maxidx, WT &minVal, WT &maxVal,
+                size_t &minIdx, size_t &maxIdx, const WT minInit, const WT maxInit,
+                const int nlanes, int len, size_t startidx, int &j, int &len0 )
+{
+    len0 = len & -nlanes;
+    j = 0;
+
+    minVal = *minval, maxVal = *maxval;
+    minIdx = *minidx, maxIdx = *maxidx;
+
+    // To handle start values out of range
+    if ( maxVal < minInit || minVal > maxInit )
+    {
+        uchar done = 0x00;
+
+        for ( ; (j < len) && (done != 0x03); j++ )
+        {
+            if ( !mask || mask[j] ) {
+                T val = src[j];
+                if ( val < minVal )
+                {
+                    minVal = val;
+                    minIdx = startidx + j;
+                    done |= 0x01;
+                }
+                if ( val > maxVal )
+                {
+                    maxVal = val;
+                    maxIdx = startidx + j;
+                    done |= 0x02;
+                }
+            }
+        }
+
+        len0 = j + ((len - j) & -nlanes);
+    }
+}
+
+#if CV_SIMD128_64F
+inline double v_reduce_min(const v_float64x2& a)
+{
+    double CV_DECL_ALIGNED(32) idx[2];
+    v_store_aligned(idx, a);
+    return std::min(idx[0], idx[1]);
+}
+
+inline double v_reduce_max(const v_float64x2& a)
+{
+    double CV_DECL_ALIGNED(32) idx[2];
+    v_store_aligned(idx, a);
+    return std::max(idx[0], idx[1]);
+}
+
+#if !CV_NEON
+inline bool v_check_any(const v_uint64x2& a)
+{
+    return v_check_any(v_reinterpret_as_f64(a));
+}
+#endif
+#endif
+
+#define MINMAXIDX_REDUCE(suffix, RT, valMin, valMax, idxMin, idxMax, none, \
+                         minVal, maxVal, minIdx, maxIdx, delta, idxArr) \
+template<typename T, typename VT, typename IT, typename IR> forceinline void \
+minMaxIdx_reduce_##suffix( VT valMin, VT valMax, IT idxMin, IT idxMax, IT none, \
+                  T &minVal, T &maxVal, size_t &minIdx, size_t &maxIdx, \
+                  size_t delta, IR* idxArr ) \
+{ \
+    if ( v_check_any(idxMin != none) ) \
+    { \
+        RT vmin = v_reduce_min(valMin); \
+        v_store(idxArr, idxMin); \
+        \
+        unsigned int posMask = v_signmask(v_setall_##suffix(vmin) == valMin); \
```

Comment:
I suggest defining `VT operator ==(VT v, ST s){return v_setall_##suffix(s) == v}`. That's the only part that changes for different MACRO instantiations.

Reply:
There is also 'RT' type.  How does this help?  I still need a macro for the suffix type.

Reply:
RT could be deduced from VT as `typename VT::lane_type`. Comparison operator could be defined as template specialization for all necessary cases. Even if macro is used for that its size will be reduced essentially.